v-next/hardhat/test/internal/builtin-plugins/network-manager/json-rpc-request-modifiers/hooks-mock.ts

```diff
@@ -0,0 +1,68 @@
+import type { HookContext } from "../../../../../src/types/hooks.js";
+import type { HardhatRuntimeEnvironment } from "../../../../../src/types/hre.js";
+import type {
+  ChainType,
+  NetworkConnection,
+} from "../../../../../src/types/network.js";
+import type {
+  JsonRpcRequest,
+  JsonRpcResponse,
+} from "../../../../../src/types/providers.js";
+
+import { createHardhatRuntimeEnvironment } from "../../../../../src/hre.js";
+
+// We override the network hook handlers to set two dynamic hook handler
```

Comment:
There was a discussion in the design document on how to approach this issue, and this method seemed the most effective. We also considered injecting a mocked provider, but that would have required reusing the hook handler logic within the provider itself, creating a circular dependency, which is problematic since these tests need the hook chain to be executed.
See [design doc here](https://www.notion.so/nomicfoundation/M2-Replace-provider-wrappers-using-hooks-112578cdeaf580559b67d384bc1e767c?pvs=4#113578cdeaf580009833c72fc4b8da1f)

Reply:
I didn't know this was part of the design doc, thanks for pointing it out! However, I was suggesting injecting the `undici` dispatcher instead of a mocked provider. However, we can think of this as a future enhancement.